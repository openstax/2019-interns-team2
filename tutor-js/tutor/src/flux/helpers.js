import _ from 'underscore';
import flux from 'flux-react';

const LOADING = 'loading';
const LOADED  = 'loaded';
const FAILED  = 'failed';
const SAVING  = 'saving';
const DELETING = 'deleting';
const DELETED = 'deleted';

const STATES = {
  LOADING,
  LOADED,
  FAILED,
  SAVING,
  DELETING,
  DELETED,
};

let idCounter = 0;
const CREATE_KEY = () => `_CREATING_${idCounter++}`;

const isNew = id => /_CREATING_/.test(id);

const CrudConfig = function() {
  return {
    _asyncStatus: {},
    _local: {},
    _changed: {},
    _errors: {},
    _reload: {},

    // If the specific type needs to do something else to the object:
    // _loaded : (obj, id) ->
    // _saved : (obj, id) ->
    // _reset : ->

    reset() {
      this._asyncStatus = {};
      this._local = {};
      this._changed = {};
      this._errors = {};
      this._reload = {};

      if (typeof this._reset === 'function') {
        this._reset();
      }
      return this.emitChange();
    },

    dontReload(id) {
      return (this._asyncStatus[id] === LOADED) && this._HACK_DO_NOT_RELOAD;
    },

    FAILED(status, msg, id, ...args) {
      this._asyncStatus[id] = FAILED;
      this._errors[id] = msg;
      if (typeof this._failed === 'function') {
        this._failed({ status, msg }, id, ...Array.from(args));
      }

      if (status !== 0) { // indicates network failure
        delete this._local[id];
        return this.emitChange();
      }
    },

    load(id) {
      return this._load(id);
    },

    // allow aliasing load without triggering api call.
    _load(id) {
      // Add a shortcut for unit testing
      if (this.dontReload(id)) { return; }
      this._reload[id] = false;
      this._asyncStatus[id] = LOADING;
      return this.emitChange();
    },

    loaded(obj, id, ...args) {
      // id = obj.id
      this._asyncStatus[id] = LOADED;

      if (obj) {
        // If the specific type needs to do something else to the object:
        this._local[id] = (typeof this._loaded === 'function' ? this._loaded(obj, id, ...Array.from(args)) : undefined) || obj;
      }

      return this.emitChange();
    },

    save(id, obj) {
      // Note: id could be isNew()
      this._asyncStatus[id] = SAVING;
      return this.emitChange();
    },

    saved(result, id, ...args) {
      // id = result.id
      this._asyncStatus[id] = LOADED; // TODO: Maybe make this SAVED

      // If the specific type needs to do something else to the object:
      const obj = typeof this._saved === 'function' ? this._saved(result, id, ...Array.from(args)) : undefined;
      if (obj) { result = obj; }

      if (result) {
        this._local[id] = result;
        this._local[result.id] = result;
        delete this._changed[result.id];
      } else {
        console.warn('API WARN: Server did not return JSON after saving. Patching locally');
        // Merge all the local changes into the new local object
        this._local[id] = _.extend(this._local[id], this._changed[id]);
      }
      delete this._changed[id];
      delete this._errors[id];
      // If the specific type needs to do something else to the object:
      return this.emitChange();
    },

    create(localId, attributes = {}) {
      if (!isNew(localId)) { throw new Error('BUG: MUST provide a local id'); }
      this._local[localId] = {};
      this._changed[localId] = attributes;
      return this._asyncStatus[localId] = LOADED;
    },

    created(result, localId, ...args) {
      this._local[localId] = result; // HACK: So react component can still manipulate the same object
      this._local[result.id] = result;
      this._asyncStatus[localId] = LOADED;
      this._asyncStatus[result.id] = LOADED;
      this.emitChange();
      return (typeof this._created === 'function' ? this._created(result, localId, ...Array.from(args)) : undefined);
    },

    _change(id, obj) {
      if (this._changed[id] == null) { this._changed[id] = {}; }
      _.extend(this._changed[id], obj);
      return this.emitChange();
    },

    _save(id) {
      return this._asyncStatus[id] = SAVING;
    },

    delete(id) {
      this._asyncStatus[id] = DELETING;
      return (typeof this._delete === 'function' ? this._delete(id) : undefined);
    },

    deleted(result, id, ...args) {
      this._asyncStatus[id] = DELETED;
      delete this._local[id];
      this.emitChange();
      return (typeof this._deleted === 'function' ? this._deleted(result, id, ...Array.from(args)) : undefined);
    },

    clearChanged(id) {
      return delete this._changed[id];
    },

    HACK_DO_NOT_RELOAD(bool) { return this._HACK_DO_NOT_RELOAD = bool; },

    // Keep this here so other exports method have access to it
    _get(id) {
      const val = this._local[id];
      if (!val && (this._asyncStatus[id] !== SAVING)) { return null; }
      // Scores stores an Array unlike most other stores
      if (_.isArray(val)) {
        return val;
      } else {
        return _.extend({}, val, this._changed[id]);
      }
    },

    exports: {
      isUnknown(id) { return !this._asyncStatus[id]; },
      isLoading(id) { return this._asyncStatus[id] === LOADING; },
      isLoaded(id) { return this._asyncStatus[id] === LOADED; },
      isDeleting(id) { return this._asyncStatus[id] === DELETING; },
      isSaving(id) { return this._asyncStatus[id] === SAVING; },
      isFailed(id) { return this._asyncStatus[id] === FAILED; },
      getAsyncStatus(id) { return this._asyncStatus[id]; },
      get(id) { return this._get(id); },
      isChanged(id) { return !_.isEmpty(this._changed[id]); },
      getChanged(id) { return this._changed[id] || {}; },
      freshLocalId() { return CREATE_KEY(); },
      isNew(id) { return isNew(id); },
      reload(id) { return this._reload[id]; },
      getError(id) { return this._errors[id]; },
    },
  };
};

// Helper for creating a simple store for actions
const makeSimpleStore = function(storeConfig) {

  let actionsConfig = _.omit(storeConfig, (value, key) => (typeof value !== 'function') || (key === 'exports'));

  actionsConfig = _.keys(actionsConfig);
  const actions = flux.createActions(actionsConfig);
  storeConfig.actions = _.values(actions);
  const store = flux.createStore(storeConfig);
  return { actions, store };
};


const extendConfig = function(newConfig, origConfig) {
  if (newConfig.exports == null) { newConfig.exports = {}; }
  _.defaults(newConfig, origConfig);
  _.defaults(newConfig.exports, origConfig.exports);
  return newConfig;
};


const makeStandardStore = function(name, storeConfig) {
  extendConfig(storeConfig, new CrudConfig());
  const { actions, store } = makeSimpleStore(storeConfig);
  return { [`${name}Actions`]: actions, [`${name}Store`]: store };
};


export { CrudConfig, makeSimpleStore, extendConfig, makeStandardStore, isNew, STATES };
