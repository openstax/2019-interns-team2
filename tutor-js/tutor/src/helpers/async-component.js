import React from 'react';
import Loadable from 'react-loadable';
import ErrorComponent from '../components/error-monitoring/async-load-error';
import Loading from 'shared/components/loading-animation';

export function asyncComponent(loader, name = '') {
  const loading = ({ error }) => {
    return (
      error ? <ErrorComponent error={error} />
        : <Loading message={`Loading ${name}â€¦`} />
    );
  };
  return Loadable({
    loader, loading,
  });

}

export function loadAsync(resolve, name) {
  // return a function so the router will only evaluate it when it's needed
  // in the future we can insert role dependant logic here
  return () => asyncComponent(resolve, name);
}
