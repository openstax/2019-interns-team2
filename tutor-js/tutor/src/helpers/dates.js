import moment from 'moment';
import 'twix';
import 'moment-timezone';
import { reduce } from 'lodash';

export function getDay(oneMoment) {
  return moment(oneMoment)
    .startOf('day')
    .twix(moment(oneMoment).endOf('day'), { allDay: true });
}

export function getDurationFromMoments(listOfMoments) {
  return reduce(
    listOfMoments,
    (current, next) => current.union(getDay(next)),
    getDay(listOfMoments[0])
  );
}
