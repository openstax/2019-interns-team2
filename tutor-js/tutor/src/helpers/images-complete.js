import { isFunction } from 'lodash';

export default function imagesComplete({
  body = document.body,
  timeoutAfter = 10000, // in ms, 10 seconds
} = {}) {
  return new Promise((resolve, reject) => {
    if (!body || !isFunction(body.querySelectorAll)) {
      resolve([]);
      return;
    }

    const images = Array.from(body.querySelectorAll('img'));
    if (0 === images.length) {
      resolve(images);
      return;
    }
    let complete = 0;
    let pendingTimeout = setTimeout(() => {
      pendingTimeout = null;
      reject(images);
    }, timeoutAfter);
    const markComplete = () => {
      complete += 1;
      if (complete === images.length) {
        if (pendingTimeout) {
          clearInterval(pendingTimeout);
          pendingTimeout = null;
        }
        resolve(images);
      }
    };
    images.forEach((img) => {
      if (img.naturalWidth) {
        markComplete();
      } else {
        img.addEventListener( 'load', markComplete, false );
        img.addEventListener( 'error', markComplete, false );
      }
    });
  });

}
