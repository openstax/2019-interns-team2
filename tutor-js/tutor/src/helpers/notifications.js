import { NotificationActions } from 'shared';
import Router from './router';

export default {
  start(bootstrapData) {
    NotificationActions.startPolling();
    return (() => {
      const result = [];
      const object = bootstrapData.flash || {};
      for (let level in object) {
        const message = object[level];
        result.push(NotificationActions.display({ message, level }));
      }
      return result;
    })();
  },

  buildCallbackHandlers(comp) {
    const { router } = comp.context;
    if (!router) {
      throw new Error('Component\'s context must have router present');
    }
    return {
      missing_student_id: {
        onAdd({ course }) {
          return router.history.push(
            Router.makePathname('changeStudentId', { courseId: course.id })
          );
        },
      },
    };
  },
};
