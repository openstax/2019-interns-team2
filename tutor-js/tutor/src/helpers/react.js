import React from 'react';
import ReactDOM from 'react-dom';
import cn from 'classnames';
import { propTypes as mobxPropTypes } from 'mobx-react';
import PropTypes from 'prop-types';
import invariant from 'invariant';
import styled, { css } from 'styled-components';
import { autobind } from 'core-decorators';
import { idType } from 'shared';

export {
  React, cn, ReactDOM, mobxPropTypes, PropTypes,
  invariant, styled, css, idType, autobind,
};
export { observable, action, computed } from 'mobx';
export { observer, inject, Provider } from 'mobx-react';
