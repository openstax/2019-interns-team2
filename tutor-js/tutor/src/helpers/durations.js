import moment from 'moment';
import _ from 'underscore';
import { TimeStore } from '../flux/time';

export default {
  create(startTime, endTime) {
    return moment(startTime).twix(endTime);
  },

  isPastDue({ due_at }) {
    return moment(TimeStore.getNow()).isAfter(due_at);
  },
};
