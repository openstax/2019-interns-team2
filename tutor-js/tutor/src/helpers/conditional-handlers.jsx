import React from 'react';
import { extend } from 'lodash';
import { asyncComponent } from './async-component';
import { CourseNotFoundWarning } from '../components/course-not-found-warning';
import Courses from '../models/courses-map';
import { OXMatchByRouter } from 'shared';

const StudentDashboard = asyncComponent(
  () => import('../screens/student-dashboard')
);

const TeacherDashboard = asyncComponent(
  () => import('../screens/teacher-dashboard')
);

const getConditionalHandlers = (Router) => {
  const MatchForTutor = OXMatchByRouter(Router, null, 'TutorRouterMatch');

  const renderDashboard = (props) => {
    const { courseId } = props.params;

    extend(props, { courseId });
    const course = Courses.get(courseId);

    if (!course) {
      return <CourseNotFoundWarning />;
    }

    if (!props.match.isExact) {
      return (
        <MatchForTutor {...props} />
      );
    }

    if (course.currentRole.isTeacher) {
      return <TeacherDashboard {...props} />;
    } else {
      return <StudentDashboard {...props} />;
    }
  };

  return {
    dashboard() { return renderDashboard; },
  };
};

export { getConditionalHandlers };
