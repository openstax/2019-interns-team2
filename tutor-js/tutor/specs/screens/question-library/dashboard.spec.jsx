import { EnzymeContext, C } from '../../helpers';
import Factory, { FactoryBot } from '../../factories';
import { slice, last } from 'lodash';
import { SnapShot, Wrapper } from '../../helpers';
import Dashboard from '../../../src/screens/question-library/dashboard';
import ExerciseHelpers from '../../../src/helpers/exercise';
import ScrollTo from '../../../src/helpers/scroll-to';

jest.mock('../../../../shared/src/components/html', () => ({ html }) =>
  html ? <div dangerouslySetInnerHTML={{ __html: html }} /> : null
);
jest.mock('../../../src/helpers/exercise');
jest.mock('../../../src/helpers/scroll-to');

describe('Questions Dashboard Component', function() {
  let props, course, exercises, book, page_ids;

  beforeEach(function() {
    course = Factory.course();
    book = course.referenceBook;
    course.referenceBook.onApiRequestComplete({ data: [FactoryBot.create('Book')] });
    exercises = Factory.exercisesMap({ ecosystem_id: course.ecosystem_id, pageIds: [], count: 0 });

    exercises.fetch = jest.fn(() => Promise.resolve());
    page_ids = slice(Array.from(course.referenceBook.pages.byId.keys()), 2, 5);
    const items = page_ids.map(page_id =>
      FactoryBot.create(
        'TutorExercise',
        {
          pool_types: ['reading_dynamic'],
          page_uuid: book.pages.byId.get(page_id).uuid,
        },
      ),
    );
    exercises.onLoaded({ data: { items } }, [{ book, page_ids }]);
    props = {
      course,
      exercises,
    };
  });

  const displayExercises = () => {
    const dash = mount(<C><Dashboard {...props} /></C>);
    dash.find('.chapter-checkbox button').at(1).simulate('click');
    dash.find('.section-controls .btn-primary').simulate('click');
    return dash;
  };

  it('matches snapshot', () => {
    expect.snapshot(<C><Dashboard {...props} /></C>).toMatchSnapshot();
  });

  it('fetches and displays', () => {
    const dash = mount(<C><Dashboard {...props} /></C>);
    expect(dash).not.toHaveRendered('.no-exercises');
    dash.find(`StyledComponent[data-section-id="${page_ids[0]}"]`).simulate('click');
    dash.find('.section-controls .btn-primary').simulate('click');
    expect(exercises.fetch).toHaveBeenCalledWith({
      course: course, limit: false, page_ids: [page_ids[0]],
    });
    expect(dash).not.toHaveRendered('.no-exercises');
    dash.unmount();
  });

  it('renders exercise details', () => {
    const dash = displayExercises();
    expect(dash).not.toHaveRendered('.no-exercises');
    const uid = dash.find('[data-exercise-id]').last().prop('data-exercise-id');
    dash.find(`[data-exercise-id="${uid}"] .action.details`).simulate('click');
    expect(dash).toHaveRendered('.exercise-details');
    dash.find('a.show-cards').simulate('click');
    expect(dash).toHaveRendered('.exercise-cards');
    expect(last(ScrollTo.mock.instances).scrollToSelector).toHaveBeenCalledWith(
      `[data-exercise-id="${uid}"]`, { immediate: true }
    );
    dash.unmount();
  });

  it('can exclude exercises', () => {
    const dash = displayExercises();
    expect(dash).not.toHaveRendered('.no-exercises');
    dash.find('.action.details').at(0).simulate('click');
    expect(dash).toHaveRendered('.exercise-details');
    course.saveExerciseExclusion = jest.fn();
    const uid = dash.find('[data-exercise-id]').last().prop('data-exercise-id');
    const exercise = exercises.array.find(e => uid == e.content.uid);
    dash.find(`[data-exercise-id="${uid}"] .action.exclude`).simulate('click');
    expect(course.saveExerciseExclusion).toHaveBeenCalledWith({ exercise, is_excluded: true });
    dash.unmount();
  });

  it('can report errors', () => {
    const dash = displayExercises();
    expect(dash).not.toHaveRendered('.no-exercises');
    dash.find('.action.details').at(0).simulate('click');
    dash.find('.action.report-error').simulate('click');
    const uid = dash.find('[data-exercise-id]').last().prop('data-exercise-id');
    const exercise = exercises.array.find(e => uid == e.content.uid);
    expect(ExerciseHelpers.openReportErrorPage).toHaveBeenCalledWith(exercise, course);
    dash.unmount();
  });

  it('clears when cancel is clicked', () => {
    const dash = mount(<C><Dashboard {...props} /></C>);
    dash.find('.chapter-checkbox button').at(1).simulate('click');
    expect(dash.find('.section.selected')).toHaveLength(9);
    dash.find('button.cancel').simulate('click');
    expect(dash.find('.section.selected')).toHaveLength(0);
  });

});
