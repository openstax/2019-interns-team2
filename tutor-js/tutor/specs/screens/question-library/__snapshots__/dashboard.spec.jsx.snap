// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Questions Dashboard Component matches snapshot 1`] = `
.c3 {
  margin-right: 0.5rem;
  margin-left: 0.5rem;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: 0.5rem;
  cursor: pointer;
  font-size: 1.5rem;
  border-bottom: 1px solid #f1f1f1;
}

.c4 > * {
  margin-left: 1rem;
}

.c4 input {
  font-size: 1.7rem;
  margin-left: 1.2rem;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 3.5rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  font-size: 1.7rem;
  border-bottom: 1px solid #f1f1f1;
}

.c2 > * {
  margin-left: 1rem;
}

.c2 .chapter-title {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 .browse-the-book {
  font-size: 1.4rem;
  margin-right: 2rem;
}

.c0 {
  border: 1px solid #f1f1f1;
  border-radius: 4px;
}

<div
  className="openstax-debug-content"
>
  <div
    data-purpose="tour-conductor-wrapper"
  >
    <div>
      <div
        className="questions-dashboard"
      >
        <div
          className="sections-chooser panel"
        >
          <div
            className="header"
          >
            <div
              className="wrapper"
            >
              <h2>
                Question Library
              </h2>
              <a
                className="btn btn-default"
                href="/course/1"
                onClick={[Function]}
              >
                Back to Dashboard
              </a>
            </div>
          </div>
          <div
            className="sections-list"
            data-tour-region-id="question-library-sections-chooser"
          >
            <div
              className="c0"
            >
              <div
                className="chapter c1"
                data-is-expanded={true}
              >
                <div
                  className="c2"
                  data-chapter-section={1}
                  onClick={[Function]}
                  role="button"
                >
                  <span
                    className="chapter-checkbox"
                  >
                    <span
                      className="tri-state-checkbox unchecked"
                      onClick={[Function]}
                      tabIndex={1}
                    >
                      <button
                        className="ox-icon-square btn btn-plain"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          className="svg-inline--fa fa-square fa-w-14 ox-icon ox-icon-square c3"
                          color="#929292"
                          data-icon="square"
                          data-prefix="far"
                          focusable="false"
                          role="img"
                          style={
                            Object {
                              "cursor": "pointer",
                            }
                          }
                          viewBox="0 0 448 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-6 400H54c-3.3 0-6-2.7-6-6V86c0-3.3 2.7-6 6-6h340c3.3 0 6 2.7 6 6v340c0 3.3-2.7 6-6 6z"
                            fill="currentColor"
                            style={Object {}}
                          />
                        </svg>
                      </button>
                    </span>
                  </span>
                  <span
                    className="chapter-number"
                  >
                    Chapter 
                    <span
                      className="chapter-section"
                      data-chapter-section="1.0"
                    >
                      1
                    </span>
                     - 
                  </span>
                  <span
                    className="chapter-title"
                  >
                     
                    Kinematics
                     
                  </span>
                </div>
                <div
                  aria-expanded={null}
                  className="sections collapse show"
                >
                  <div
                    className="section c4"
                    data-section-id="234"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="chapter-section"
                      data-chapter-section="1.1"
                    >
                      1.1
                    </span>
                    <span
                      className="section-title"
                    >
                       
                      Kinematics in 1 dim
                    </span>
                  </div>
                  <div
                    className="section c4"
                    data-section-id="235"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="chapter-section"
                      data-chapter-section="1.2"
                    >
                      1.2
                    </span>
                    <span
                      className="section-title"
                    >
                       
                      Kinematics in 2 dim
                    </span>
                  </div>
                  <div
                    className="section c4"
                    data-section-id="236"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="chapter-section"
                      data-chapter-section="1.3"
                    >
                      1.3
                    </span>
                    <span
                      className="section-title"
                    >
                       
                      Kinematics in 3 dim
                    </span>
                  </div>
                </div>
              </div>
              <div
                className="chapter c1"
                data-is-expanded={false}
              >
                <div
                  className="c2"
                  data-chapter-section={2}
                  onClick={[Function]}
                  role="button"
                >
                  <span
                    className="chapter-checkbox"
                  >
                    <span
                      className="tri-state-checkbox unchecked"
                      onClick={[Function]}
                      tabIndex={1}
                    >
                      <button
                        className="ox-icon-square btn btn-plain"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          className="svg-inline--fa fa-square fa-w-14 ox-icon ox-icon-square c3"
                          color="#929292"
                          data-icon="square"
                          data-prefix="far"
                          focusable="false"
                          role="img"
                          style={
                            Object {
                              "cursor": "pointer",
                            }
                          }
                          viewBox="0 0 448 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-6 400H54c-3.3 0-6-2.7-6-6V86c0-3.3 2.7-6 6-6h340c3.3 0 6 2.7 6 6v340c0 3.3-2.7 6-6 6z"
                            fill="currentColor"
                            style={Object {}}
                          />
                        </svg>
                      </button>
                    </span>
                  </span>
                  <span
                    className="chapter-number"
                  >
                    Chapter 
                    <span
                      className="chapter-section"
                      data-chapter-section="2.0"
                    >
                      2
                    </span>
                     - 
                  </span>
                  <span
                    className="chapter-title"
                  >
                     
                    Mechanics
                     
                  </span>
                </div>
                <div
                  aria-expanded={null}
                  className="sections collapse"
                >
                  <div
                    className="section c4"
                    data-section-id="237"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="chapter-section"
                      data-chapter-section="2.1"
                    >
                      2.1
                    </span>
                    <span
                      className="section-title"
                    >
                       
                      Mechanics in 1 dim
                    </span>
                  </div>
                  <div
                    className="section c4"
                    data-section-id="238"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="chapter-section"
                      data-chapter-section="2.2"
                    >
                      2.2
                    </span>
                    <span
                      className="section-title"
                    >
                       
                      Mechanics in 2 dim
                    </span>
                  </div>
                  <div
                    className="section c4"
                    data-section-id="239"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="chapter-section"
                      data-chapter-section="2.3"
                    >
                      2.3
                    </span>
                    <span
                      className="section-title"
                    >
                       
                      Mechanics in 3 dim
                    </span>
                  </div>
                </div>
              </div>
              <div
                className="chapter c1"
                data-is-expanded={false}
              >
                <div
                  className="c2"
                  data-chapter-section={3}
                  onClick={[Function]}
                  role="button"
                >
                  <span
                    className="chapter-checkbox"
                  >
                    <span
                      className="tri-state-checkbox unchecked"
                      onClick={[Function]}
                      tabIndex={1}
                    >
                      <button
                        className="ox-icon-square btn btn-plain"
                        disabled={false}
                        onClick={[Function]}
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          className="svg-inline--fa fa-square fa-w-14 ox-icon ox-icon-square c3"
                          color="#929292"
                          data-icon="square"
                          data-prefix="far"
                          focusable="false"
                          role="img"
                          style={
                            Object {
                              "cursor": "pointer",
                            }
                          }
                          viewBox="0 0 448 512"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M400 32H48C21.5 32 0 53.5 0 80v352c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48V80c0-26.5-21.5-48-48-48zm-6 400H54c-3.3 0-6-2.7-6-6V86c0-3.3 2.7-6 6-6h340c3.3 0 6 2.7 6 6v340c0 3.3-2.7 6-6 6z"
                            fill="currentColor"
                            style={Object {}}
                          />
                        </svg>
                      </button>
                    </span>
                  </span>
                  <span
                    className="chapter-number"
                  >
                    Chapter 
                    <span
                      className="chapter-section"
                      data-chapter-section="3.0"
                    >
                      3
                    </span>
                     - 
                  </span>
                  <span
                    className="chapter-title"
                  >
                     
                    Thermodynamics
                     
                  </span>
                </div>
                <div
                  aria-expanded={null}
                  className="sections collapse"
                >
                  <div
                    className="section c4"
                    data-section-id="240"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="chapter-section"
                      data-chapter-section="3.1"
                    >
                      3.1
                    </span>
                    <span
                      className="section-title"
                    >
                       
                      Thermodynamics in 1 dim
                    </span>
                  </div>
                  <div
                    className="section c4"
                    data-section-id="241"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="chapter-section"
                      data-chapter-section="3.2"
                    >
                      3.2
                    </span>
                    <span
                      className="section-title"
                    >
                       
                      Thermodynamics in 2 dim
                    </span>
                  </div>
                  <div
                    className="section c4"
                    data-section-id="242"
                    onClick={[Function]}
                  >
                    <span
                      className="section-checkbox"
                    >
                      <input
                        checked={false}
                        readOnly={true}
                        type="checkbox"
                      />
                    </span>
                    <span
                      className="chapter-section"
                      data-chapter-section="3.2"
                    >
                      3.2
                    </span>
                    <span
                      className="section-title"
                    >
                       
                      Thermodynamics in 3 dim
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            className="section-controls footer"
          >
            <div
              className="wrapper"
            >
              <button
                className="btn btn-primary"
                disabled={true}
                onClick={[Function]}
                type="button"
              >
                Show Questions
              </button>
              <button
                className="cancel btn btn-default"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                Cancel
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="visible-when-debugging"
    >
      <div
        className="tour-spy-info"
      >
        available regions: []; region tour ids: []; audience tags: []; tour tags: []; eligible tours: []; TOUR RIDE: &lt;none&gt;
      </div>
    </div>
  </div>
  <a
    aria-hidden="true"
    className="debug-toggle-link"
    href="#spy"
    onClick={[Function]}
    tabIndex={-1}
  >
    π
  </a>
</div>
`;
